pkg/client/kubernetes/clientmap/internal/garden_clientmap.go

```diff
@@ -52,18 +65,64 @@ func (f *GardenClientSetFactory) CalculateClientSetHash(context.Context, clientm
 }
 
 // NewClientSet creates a new ClientSet to the garden cluster.
-func (f *GardenClientSetFactory) NewClientSet(_ context.Context, k clientmap.ClientSetKey) (kubernetes.Interface, error) {
+func (f *GardenClientSetFactory) NewClientSet(ctx context.Context, k clientmap.ClientSetKey) (kubernetes.Interface, error) {
 	_, ok := k.(GardenClientSetKey)
 	if !ok {
 		return nil, fmt.Errorf("unsupported ClientSetKey: expected %T got %T", GardenClientSetKey{}, k)
 	}
 
-	return NewClientSetWithConfig(
+	configFns := []kubernetes.ConfigFunc{
 		kubernetes.WithRESTConfig(f.RESTConfig),
 		kubernetes.WithClientOptions(client.Options{
 			Scheme: kubernetes.GardenScheme,
 		}),
-	)
+	}
+
+	// Create multi-namespace cache for secrets
+	var seedNamespaces []string
+	if seedName := f.SeedSelector.SeedName; len(seedName) > 0 {
+		seedNamespaces = append(seedNamespaces, seedpkg.ComputeGardenNamespace(seedName))
+	}
+	if ls := f.SeedSelector.Selector; ls != nil {
+		directClient, err := client.New(f.RESTConfig, client.Options{
+			Scheme: kubernetes.GardenScheme,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		seeds := &corev1beta1.SeedList{}
+
+		seedSelector, err := metav1.LabelSelectorAsSelector(ls)
+		if err != nil {
+			return nil, err
+		}
+
+		if err := directClient.List(ctx, seeds, client.MatchingLabelsSelector{Selector: seedSelector}); err != nil {
+			return nil, err
+		}
+
+		for _, seed := range seeds.Items {
+			seedNamespaces = append(seedNamespaces, seedpkg.ComputeGardenNamespace(seed.Name))
```

Comment:
What happens when a new seed that would match would be added? I guess this might need to be refreshed when the client is retrieved from the client map, or?

Edit: Ah, or is this done here (https://github.com/gardener/gardener/pull/3700/commits/007c16d467a077aea8976410e7c5426388185e6d#diff-628c8891682f4a4b0f5c49cb8051057ed8a5a3bf338438659ef68ccfb66385e5R99-R103)? But then it must be done for all the places where the garden client is used and interacts with only matching seeds, or?

Reply:
Yes, the seed controller (pkg/gardenlet/controller/seed/seed_control.go) will invalidate the Garden client when (a) a new seed is added or (b) a seed is removed because then we expect new/removed namespaces in the Garden cluster. 

> But then it must be done for all the places where the garden client is used and interacts with only matching seeds, or?

As soon as the Garden client is requested again  in one of the many places (after it was invalidated), the updated seed list (either with the new or removed seed) is considered. Not sure if you have another use-case in mind?

Reply:
I think @timebertt had implemented a mechanism that allows to invalidate/refresh the client automatically so that you don't have to handle it explicitly in the "business logic" code. For example, for the `Shoot` client a hash is computed and if it changes then it's auto-refreshed, see https://github.com/gardener/gardener/blob/4c6153dceae2e461607c9a38afca88277387dda0/pkg/client/kubernetes/clientmap/internal/shoot_clientmap.go#L62-L80

The garden clientmap doesn't implement this yet, see https://github.com/gardener/gardener/blob/4c6153dceae2e461607c9a38afca88277387dda0/pkg/client/kubernetes/clientmap/internal/garden_clientmap.go#L49-L52

Maybe we can use this? Probably @timebertt can help and guide us here a little :)

Reply:
> I think @timebertt had implemented a mechanism that allows to invalidate/refresh the client automatically so that you don't have to handle it explicitly in the "business logic" code. For example, for the Shoot client a hash is computed and if it changes then it's auto-refreshed, see

Yep, this is exactly how `CalculateClientSetHash` is supposed to be used: to implicitly refresh clientsets if the underlying configuration changed.

Though, I'm asking myself if this special handling for the `seedSelector` use case (i.e. gardenlet running outside of the seed cluster) is really worth the effort and maintenance complexity. This will likely cause us some headache in the future. 
WDYT not doing any special cache handling for a gardenlet responsible for multiple seeds (i.e. configuring a `seedSelector`). This use case is anyways not supposed to be used in production and has probably also other security flaws. If we just don't do those cache optimizations for the "dev scenario" it will save us from refreshing the clientsets.

Reply:
Hm, how much code would it be? I'm not too happy with intentionally not fully supporting the configuration options we offer.

Generally, maybe we should then rather go into the direction of dropping support for this `seedSelector` feature if we anyways don't recommend it for productive setups and don't know of use-cases (other than dev). Assuming it's dropped, we can make the dev environment work by simply not providing any `seedConfig` which would mean "target all seeds".

Reply:
> Generally, maybe we should then rather go into the direction of dropping support for this seedSelector feature if we anyways don't recommend it for productive setups and don't know of use-cases (other than dev). Assuming it's dropped, we can make the dev environment work by simply not providing any seedConfig which would mean "target all seeds".

That sounds like a good idea to me.

Reply:
> Assuming it's dropped, we can make the dev environment work by simply not providing any seedConfig which would mean "target all seeds".

Hm, especially for dev uses cases one adds/removes seeds and therefore you need to consider newly added or removed seed namespaces in the Garden cluster. Here it doesn't make a difference if you target a subset of seeds or all seeds.

Reply:
Hm, that's true. @timebertt would you still vote for not supporting it? Personally, I don't feel too comfortable with this approach...

Reply:
> Here it doesn't make a difference if you target a subset of seeds or all seeds.

That's true.

What I actually meant with 

> WDYT not doing any special cache handling for a gardenlet responsible for multiple seeds (i.e. configuring a seedSelector). This use case is anyways not supposed to be used in production and has probably also other security flaws. If we just don't do those cache optimizations for the "dev scenario" it will save us from refreshing the clientsets.

is to basically don't use any caching at all for secrets if the gardenlet is responsible for mulitple seeds. We could simply always read directly from the garden cluster and thereby avoid clientset/cache invalidating/rebuilding.

I'm not insisting on it, just trying to figure out, which way we are the least likely to run into bugs in the future because of added complexity.
WDYT?

Reply:
Ah, interesting, I didn't get it that way. But yeah, I think that would be totally fine. However, I cannot judge what this means with respect to the code complexity. @timuthy :)

Reply:
I recognized another problem with the approach to dynamically invalidate the garden client during runtime.
Currently all of our controllers assume, that the garden clientset is never invalidated (the watches will not be reestablished when they are stopped). Meaning, if we would either need to adapt all controllers to reestablish watches or restart the gardenlet if the garden client should be invalidated...
Neither of this seems like a good option to me.

Reply:
Valid points @timebertt and @rfranzke. I originally wanted to fully support cached secrets for dev/multiple seed use cases, because in production with a Gardenlet per seed, caching will become the default option and if we now workaround it during dev, bugs for common setups might remain unnoticed for a longer time. OTOH, if we don't support invalidating the Gardenclient, we should probably go with @timebertt's suggestion and skip coaching, which is a tradeoff but ATM I don't have any further ideas either.

Reply:
> because in production with a Gardenlet per seed, caching will become the default option and if we now workaround it during dev, bugs for common setups might remain unnoticed for a longer time

Yeah agree, I was also worried about this.
Unfortunately I don't have a better idea on how to deal with this 😕